### PR TITLE
Shared randomizer

### PR DESCRIPTION
**Pull request by ryley-o**
## Description of the change

Update randomizer to be shared, as well as use a separate atomic pseudorandom contract for atomic hash generation.

Relevant internal design document discussion: https://docs.google.com/document/d/1Ix0LhPHqeDgCouKPSYZufQY9zQ5cNqf8yXHbus8y3w8/edit?disco=AAAAvQ9BYpk

## Rollout
Contracts may begin using this if desired, but any contract using this randomizer will NOT be compatible with MinterPolyptychV0, and will have to wait until the minter suite re-architecture work is completed before polyptych minting will be available.

Engine deployment scripts are not updated in this PR to enable immediate use of the new shared randomizer (this is open to be changed per feedback during PR review)

## Tests

- Tests added to reach full coverage of the new contracts
- Tests written to go beyond coverage (e.g. check for events, etc.)

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204434079486934

### CODE REVIEW

**Comment by jakerockland** [comment]
@ryley-o My apology for the delay here, is this one review-ready?

**Comment by ryley-o** [comment]
>is this one review-ready?

Yup! no worries on the delay - based on the state of the design doc review, I believe this PR is ready for review, as well as #617.

**Comment by jakerockland** [comment]
> Engine deployment scripts are not updated in this PR to enable immediate use of the new shared randomizer (this is open to be changed per feedback during PR review)

SGTM to wait to do this until after minter suite rearchitecture is complete. Will note that we should include information/documentation on how to migrate to this new randomizer setup for partners in our broader migration runbook/plan for Engine partners for the minter suite migration.

File: contracts/randomizer/primitive/PseudorandomAtomicBasic.sol [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+pragma solidity 0.8.17;
+
+import "../../interfaces/0.8.x/IPseudorandomAtomic.sol";
+
+/**
+ * @title PseudorandomAtomic
+ * @author Art Blocks Inc.
+ * @notice This contract atomically returns a pseudorandom value.
+ */
+contract PseudorandomAtomic {
```
Comment (jakerockland) [inline contracts/randomizer/primitive/PseudorandomAtomicBasic.sol:13]: This singleton holds no state – may want to consider making this an externally linked library for deployment purposes (can discuss further offline if helpful)?

**Review by jakerockland — APPROVED** [review]
This generally LGTM w/ couple minor open questions @ryley-o – pending added test coverage 🙏 💜

File: contracts/randomizer/primitive/PseudorandomAtomicBasic.sol (hunk 1 above)
Comment (ryley-o) [reply contracts/randomizer/primitive/PseudorandomAtomicBasic.sol:13]: We want to have the ability to hold state in this point of our overall contract architecture 👍

File: contracts/randomizer/shared/SharedRandomizerV0.sol [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+     * with the pseudorandom number request.
+     */
+    function _getPseudorandomAtomic(
+        address _coreContract,
+        uint256 _tokenId
+    ) internal view returns (bytes32) {
+        return
+            IPseudorandomAtomic(msg.sender).getPseudorandomAtomic(
+                keccak256(abi.encodePacked(_coreContract, _tokenId))
+            );
```
[the remark this answers is not in the document]
Comment (ryley-o) [reply contracts/randomizer/shared/SharedRandomizerV0.sol:243]: yes definitely 😅 good catch

**Comment by ryley-o** [comment]
>Will note that we should include information/documentation on how to migrate to this new randomizer setup for partners in our broader migration runbook/plan for Engine partners for the minter suite migration.

Good callout, and I added a point about this in our design document 🙏

File: contracts/randomizer/primitive/PseudorandomAtomicBasic.sol (hunk 1 above)
Comment (jakerockland) [reply contracts/randomizer/primitive/PseudorandomAtomicBasic.sol:13]: 🙏 👍 SGTM

**Comment by jakerockland** [comment]
@lyaunzbe gentle ping here on additional review pass 🙏

**Review by lyaunzbe — APPROVED** [review]
lgtm 🔥

**Pull request merged**